lib/pure/collections/sequtils.nim

```diff
@@ -918,6 +1301,270 @@ template newSeqWith*(len: int, init: untyped): untyped =
+    ##
+    ## The template injects the ``it`` variable representing the previous value which
+    ## you can use directly in an expression.
+    ##
+    block:
+      var result: typeof(def) = def
+      for it {.inject.} in s:
+        if op:
+          result = it
+          break
+      result
+
+  template findItFirst*(s: typed, op: untyped): untyped =
+    ## Find the first element in `s` on which `op` evaluated to `true`. If such
+    ## element is not found is `s`, `default(T)` is returned, where `T` is the
+    ## type of items in `s`.
+    ##
+    ## The template injects the ``it`` variable representing the previous value which
+    ## you can use directly in an expression.
+    ##
+    findItFirst(s, op, default(typeof(items(s), typeOfIter)))
+
+  template indexOfIt*(s: typed, op: untyped): int =
+    ## Get the index of the first element in `s` that fulfilled the predicate
+    ## `op` (a `bool` expression). If such element is not found is `s`,
+    ##  `low(s) - 1` is returned.
+    ##
+    ## The template injects the ``it`` variable representing the previous value which
+    ## you can use directly in an expression.
+    ##
+    block:
+      evalOnceAs(s1, s, compiles((let _ = s)))
+      var result = low(s1) - 1
+      for i in low(s1) .. high(s1):
+        let it {.inject.} = s1[i]
+        if op:
+          result = i
+          break
+      result
+
+  template scanIt*(s: typed, op: untyped) =
+    ## Evaluate `op` on every item in the container `s`.
+    ##
+    ## The template injects the ``it`` variable representing the previous value which
+    ## you can use directly in an expression.
+    ##
+    runnableExamples:
+      @[1, 2, 3, 4].scanIt(echo it)
+
+    for it {.inject.} in s:
+      op
+
+  template zipThem*(s1, s2: typed; op: untyped): untyped =
+    ## Returns a new sequence with a transformation (`op`) on each pair of
+    ## elements from the two input containers.
+    ##
+    ## The template injects the ``it`` and ``jt`` variables representing the
+    ## previous values from ``s1`` and ``s2`` respectively which you can use
+    ## directly in an expression.
+    ##
+    ## The input containers can be of different types.
+    ## If one container is shorter, the remaining items in the longer container
+    ## are discarded.
+    ##
+    runnableExamples:
+      let
+        s1 = @[1, 2, 3]
+        s2 = @["I", "II", "III", "IV"]
+      assert zipThem(s1, s2, $it & ":" & jt) == @["1:I", "2:II", "3:III"]
+
+    block:
+      evalOnceAs(ss1, s1, compiles((let _ = s1)))
+      evalOnceAs(ss2, s2, compiles((let _ = s2)))
+
+      type
+        OutType = typeof((block:
+          var
+            it {.inject.}: typeof(items(ss1), typeOfIter)
+            jt {.inject.}: typeof(items(ss2), typeOfIter)
+          op), typeOfProc)
+
+      let m = min(ss1.len, ss2.len)
+      var result = newSeq[OutType](m)
+      for i in 0 ..< m:
+        let it {.inject.} = ss1[i]
+        let jt {.inject.} = ss2[i]
```

Comment:
For consistency sake, do we want to add this naming scheme of `it`, `jt`, `kt` and so on when [`fold`](https://github.com/nim-lang/Nim/blob/version-1-0/lib/pure/collections/sequtils.nim#L727), for example, is injecting `a` and `b`?

Reply:
`it`, `jt` and `kt` represents items from s1/s2/s3 respectively. They play the same role. While in `fold`, `a` is the accumulated result, `b` is an item, they play different role. Are `a` and `b` good
choices for injected variables in `fold`? This is out of the scope of this PR.